src/v/archival/ntp_archiver_service.cc

```diff
@@ -295,28 +295,68 @@ ss::future<cloud_storage::download_result> ntp_archiver::sync_manifest() {
         vlog(
           _rtclog.debug, "Updating the archival_meta_stm in read-replica mode");
         std::vector<cloud_storage::segment_meta> mdiff;
-        auto offset
-          = _partition->archival_meta_stm()->manifest().get_last_offset()
-            + model::offset(1);
-        // TODO: this code needs to be updated when the compacted segment
-        // uploads are merged.
+        // Several things has to be done:
+        // - Add all segments between old last_offset and new last_offset
+        // - Compare all segments below last compacted offset with their
+        //   counterparts in the old manifest and re-add them if they are
+        //   diferent.
+        // - Apply new start_offset if it's different
+        auto offset = model::next_offset(manifest().get_last_offset());
         for (auto it = m.segment_containing(offset); it != m.end(); it++) {
             mdiff.push_back(it->second);
         }
+
+        bool needs_cleanup = false;
+        auto old_start_offset = manifest().get_start_offset();
+        auto new_start_offset = m.get_start_offset();
+        for (const auto& s : m) {
+            if (
+              s.second.committed_offset
+                <= m.get_last_uploaded_compacted_offset()
+              && s.second.base_offset >= new_start_offset) {
+                auto iter = manifest().get(s.first);
```

Comment:
Is there a possibility that on a different leader compacted segments can have slightly different offset ranges than the original ones on the previous leader?

Reply:
👍 a comment might be helpful that the archiver guarantees it.